rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,823 @@
+          ],
+          "minimum": 1,
+          "maximum": 4294967296
+        },
+        "max_timeout": {
+          "description": "Maximum period of time a batch can exist before being forcibly flushed.",
+          "anyOf": [
+            { "type": "null" },
+            { "$ref": "#/definitions/duration" }
+          ],
+          "minimum": 1
+        }
+      }
+    },
+    "duration": {
+      "type": "number",
+      "minimum": 0,
+      "maximum": 9007199254740991
+    }
+  }
+}
+```
+
+While this is long and verbose, users almost exclusively interact with JSONSchema documents by using
+a library that validates JSON documents using the schema.  Briefly, though, you can observe the
+following things:
+
+- we can see that for the `SinkConfig::endpoint` field, we've enumerated its alias, `url`, along
+  with the fact that it's deprecated
+- throughout the various fields, we can set field constraints such as the minimum or maximum value
+  for integers
+- we've utilized a common definition of the `BatchConfig` type, but we can also do a union of its
+  definition along with a separate definition that enforces a maximum size of 1000 events per batch
+- basic information such as field and type descriptions are present, enriching the schema
+- we have support for custom metadata, which is letting us include the sink's beta status in the
+  schema for rich, semantic descriptions
+- while it does not affect validation at all, we can also include other semantic information, such
+  as the default value for a field, even when that field has a shared definition, like `BatchConfig`
+
+## Rationale
+
+In general, the lack of a configuration schema, and being able to treat our Rust code as the source
+of truth, hurts both users and developers. As such, doing this work represents a huge opportunity to
+reduce developer toil when it comes to generating and synchronizing our generated documentation.  It
+additionally represents a large quality of life improvement for our users, who look to our
+documentation to be up-to-date and semantically meaningful.
+
+If we didn't do this work, we would be stuck with the current manual toil, which is not only extra
+work on the part of developers -- learning Cue, remembering to update it, etc -- but also reviewers,
+who now need to catch when these things are missed.
+
+Utilizing an extensible system for expressing the schema of Vector's configuration gives us the
+runway to solve our current problems, but also the ability to handle future problems and
+requirements as they come up.
+
+## Drawbacks
+
+The primary drawback of this approach is that it limits us to things we can reasonably express
+within the limitations of the Rust language itself, and what we're comfortable with representing via
+attributes. For example, single-line and multi-line doc comments are trivially supported, but if we
+wanted to start pushing more semantically-relevant information, such as configuration snippets, etc,
```

Comment:
This is a big question for me, I was experimenting with examples like the following based on the Cue data we have now which I think is extremely helpful https://a.cl.ly/xQuq45LO, which is essentially an example config, input and expected output—would this be easy enough with the proposed approach?

Reply:
Also where do we plan to store longer form content associated with each component?

Reply:
I had a similar comment [here](https://github.com/vectordotdev/vector/pull/11634#pullrequestreview-896850707). I think...

1. Configuration examples should live in the `config` directory. We could have individual component examples as well as broad use case examples (i.e., cutting AWS VPC logs cost). Then use these files directly on the site.
2. Things like the "How it works" sections on the current docs would be in website markdown files somewhere. My hope is that all markdown lives in the website repo.

Open to more discussion on this, but you get the drift. I don't think the Rust code will drive _everything_, just the schema.

Reply:
Ok yes, that's addressing the same question

Reply:
Sounds like we have consensus here that we can handle the corner cases with other means.